File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
       Map<Character,Integer> map=new HashMap<>();
       int max=0;
       int pre=0;
       if(s.length()<=1) return s.length();
       map.put(s.charAt(0),0);
       for(int i=1;i<s.length();i++){
        while(map.containsKey(s.charAt(i))){
            max=Math.max(map.size(),max);
            map.remove(s.charAt(pre));
            pre++;
        }
        map.put(s.charAt(i),i);
            max=Math.max(map.size(),max);
       } 
       return max;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        map=set()

        ans=0
        j=0
        for i in range(len(s)):
            ch=s[i]
            if(ch in map):
                ans=max(ans,len(map))
                while(ch in map):
                    map.remove(s[j])
                    j+=1
                map.add(ch)
            else:
                map.add(ch)
                ans=max(ans,len(map))
        return ans
